src/ckb/channel.rs

```diff
@@ -2917,6 +2973,123 @@ impl ChannelActorState {
         Ok(())
     }
 
+    fn handle_reestablish_channel_message(
+        &mut self,
+        reestablish_channel: ReestablishChannel,
+        network: &ActorRef<NetworkActorMessage>,
+    ) -> ProcessingChannelResult {
+        debug!(
+            "Handling reestablish channel message: {:?}, our commitment_numbers {:?}",
+            reestablish_channel, self.commitment_numbers,
+        );
+        self.reestablishing = false;
+        match self.state {
+            ChannelState::NegotiatingFunding(_flags) => {
+                // TODO: in current implementation, we don't store the channel when we are in NegotiatingFunding state.
+                // This is an unreachable state for reestablish channel message. we may need to handle this case in the future.
+            }
+            ChannelState::ChannelReady() => {
+                let expected_local_commitment_number = self.get_local_commitment_number();
+                let acutal_local_commitment_number = reestablish_channel.remote_commitment_number;
+                if acutal_local_commitment_number == expected_local_commitment_number {
+                    // resend AddTlc, RemoveTlc and CommitmentSigned messages if needed
+                    let mut need_resend_commitment_signed = false;
+                    for info in self.tlcs.values() {
+                        if info.is_offered() {
```

Comment:
Here we have multiple implicit assumptions.
1. If a TLC is added by us or a TLC is removed by us then we need to resend a `CommitmentSigned` message.
2. We only need to resend a single `CommitmentSigned` message (or we only have a single TLC added/removed in this commitment phase).
3. We only need to resend `AddTlc` or `RemoveTlc` messages when the channel is established.

Assumption 1 and assumption 2 are correct in current implementation. Assumption 2 is the direct corollary of assumption 1, but the code here can be a little fragile if we don't know all these assumptions and try to refactor the logic here. Can we add a warning if there are multiple `AddTlc`/`RemoveTlc` need to be processed?

Assumption 3 is actually wrong. For `RemoveTlc`, we may need to transition the state to `Shutdown` if both parties agreed to shutdown and this is the last pending TLC to be removed. https://github.com/nervosnetwork/cfn-node/blob/975f25668aff31b54ccad0eaf97a177069367de2/src/ckb/channel.rs#L593

In regard to assumption 3, I think we can create a `AddTlc` or `RemoveTlc` from the `DetailedTlcInfo`, then call `handle_add_tlc_command` or `handle_remove_tlc_command`, and make sure `handle_add_tlc_command` and `handle_remove_tlc_command` are always idempotent. This way even if code changes, we can keep the the processing logic of `RemoveTlc` in channel reestablishment and normal operations in sync.

Reply:
We don't have to reuse all the code of `handle_add_tlc_command` and `handle_remove_tlc_command`. The common logic can be extracted to new functions.